### ISSUE TITLE
[FEA] Better Pandas GroupyBy functionality

### ISSUE THREAD

**Issue by quasiben** [rapidsai/cudf#1720]
In trying to establish better `DataFrameGroupBy`/`SeriesGroupBy` object handling in dask, `cuDF` should implement some of the following (ref: https://pandas.pydata.org/pandas-docs/stable/reference/groupby.html)

- `ngroup`
- `ngroups`
- `get_group`
- `group`

We can achieve better object detection by inspecting if these methods exist on an object, similar to how dask currently checks if the object `is dataframe like`:
https://github.com/dask/dask/blob/25adf164ee1948388c2495a3000dd730fe0d0690/dask/dataframe/utils.py#L503-L510

Additionally, the above is useful because these methods are important during standard analyses where grouping is employed

**Comment by kkraus14** [comment]
@quasiben Some of these aren't ideal to run with a hash based groupby as they're computed while you compute an aggregation which builds the hash table. It would be good to understand how / where Dask and/or other libraries use these attributes / functions for us to plan how we can build something both compatible and GPU friendly.

If you have any examples it would be greatly appreciated!

**Comment by quasiben** [comment]
`is_dataframe_like` looks like the following:

```python
def is_dataframe_like(df):
    """ Looks like a Pandas DataFrame """
    typ = type(df)
    return (all(hasattr(typ, name)
                for name in ('groupby', 'head', 'merge', 'mean')) and
            all(hasattr(df, name) for name in ('dtypes',)) and not
            any(hasattr(typ, name)
                for name in ('value_counts', 'dtype')))
```

Essentially, dask looks at some degree of similarity between the APIs of dataframe implementations.   Perhaps, minimally we could implement `ngroups` ?  Could this be done with `len(self._by)` ?

**Comment by kkraus14** [comment]
`len(self._by)` would return the number of columns we're grouping on, not the number of groups. The number of groups wouldn't be known until after the aggregation has finished running which is why this attribute is problematic for a hash based groupby.

**Comment by kkraus14** [comment]
If we can set the property to `None` and that would suffice for Dask I'm absolutely happy to do so.

**Comment by quasiben** [comment]
They might suffice.   @mrocklin what do you think ?

**Comment by mrocklin** [comment]
My gut reaction is to let this lie for a while.  Things work today.  If figure that we wait until more problems arise and we get more information to help us refine what this should look like.  Making conventions before having a lot of anecdotal data makes me less-than-confident that we'll get it correct.  

Even with a fair amount of information, we still had to modify `is_dataframe_like` several times before we got things right (and even now I'm not so sure).

**Comment by quasiben** [comment]
Thanks for the comment @mrocklin.  I'm going to close and take your suggestion.  @kkraus14 thank you as well for the responses

**Comment by quasiben** [comment]
To fully support https://github.com/rapidsai/dask-cudf/pull/262/files, I believe we will need `get_group`.  @kkraus14 here is an example in dask:

https://github.com/dask/dask/blob/master/dask/dataframe/shuffle.py#L545-L548

**Comment by kkraus14** [comment]
Fixed by #4346
